File: challenges/rev/n00b_Crypto.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* flag = "\x11Xxy4^|Q8\x16""2}\x1d{^?NQD\x1eLRsn?\x19[TSl\x1bls\x16?8t]TV{_";
int map[] = {3, 5, 1, 0, 2, 6, 4, 7};

int mixup(int val) {
    int new_val = 0;
    for(int i = 0; i < 8; i++) {
        new_val |= (val & 1) << map[i];
        val = val >> 1;
    }
    return new_val;
}

char* encrypt(char* input) {
    int n = strlen(input);
    char* output = malloc(n + 1);
    for(int i = 0; i < n; i++) {
        int val = input[i];
        int new_val = mixup(val);
        output[i] = new_val;
    }
    output[n] = 0;
    return output;
}


int main() {
    printf("Here is my superior encryption implemenation. I dare you to touch it.\n");
    printf("The thing you are looking for must be somewhere in the memory though.\n");
    printf("I forgot to mention, I am a n00b in crypto.\n");
    char input[100];
    printf("Enter a string you want to secure: ");
    scanf("%s", input);
    printf("Here is your secured input: %s\n", encrypt(input));
    return 0;
}

File: solutions/rev/n00b_Crypto.py
map = {0: 3, 1: 5, 2: 1, 3: 0, 4: 2, 5: 6, 6: 4, 7: 7}
invmap = {3: 0, 5: 1, 1: 2, 0: 3, 2: 4, 6: 5, 4: 6, 7: 7}


def mixup(val):
    new_val = 0
    for i in range(8):
        new_val |= (val & 1) << map[i]
        val = val >> 1
    return new_val


def unmix(val):
    new_val = 0
    for i in range(8):
        new_val |= (val & 1) << invmap[i]
        val = val >> 1
    return new_val


def encrypt(plaintext):
    ciphertext = ""
    for a in plaintext:
        val = ord(a)
        new_val = mixup(val)
        ciphertext += chr(new_val)
    return ciphertext


def decrypt(ciphertext):
    plaintext = ""
    for a in ciphertext:
        val = ord(a)
        new_val = unmix(val)
        plaintext += chr(new_val)
    return plaintext

enc = "\x11Xxy4^|Q8\x16""2}\x1d{^?NQD\x1eLRsn?\x19[TSl\x1bls\x16?8t]TV{_"
print(decrypt(enc))
